### ISSUE TITLE
[BUG] LightGBM past covariates look-ahead bias

### ISSUE THREAD

**Issue by tuomijal** [unit8co/darts#1822]
```python
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from darts.datasets import EnergyDataset
from darts.models import LightGBMModel
from darts import TimeSeries
```

## Describe the bug
I am concerned that there is a specific corner case of look-ahead bias. This seems to be the case with the following configuration:

- `model == LightGBMModel`
- `output_chunk_length < horizon`
- `past_covariates != None`

## To Reproduce

After loading some arbitraty data, we will generate two versions of `past_covariates` series: 

1) `pc` that has not been manipulated and consisting of `pc_train` and `pc_test` series
2) `pc_zeros` in which `pc_test` has been floored to zero

```python
data = EnergyDataset().load().pd_dataframe().head(1000)

y = data['total load actual'].fillna(0)
pc = data['generation biomass'].fillna(0)

length = 400
y_train, y_test = y.iloc[:length], y.iloc[length:]
pc_train, pc_test = pc.iloc[:length], pc[length:]

zeros = np.zeros(600)
zeros = pd.Series(zeros, index=pc_test.index)
pc_zeros = pd.concat([pc_train, zeros])

pc.iloc[390:410].plot(label='pc');
pc_zeros.iloc[390:410].plot(label='pc_zeros');
plt.legend();
```

    
![output_3_0](https://github.com/unit8co/darts/assets/23170686/c9943e54-3525-47e7-be38-e0fe131e95c0)

    

```python
# Convert to TimeSeries
pc = TimeSeries.from_series(pc)
y_train = TimeSeries.from_series(y_train)
pc_train = TimeSeries.from_series(pc_train)
pc_test = TimeSeries.from_series(pc_test)
pc_zeros = TimeSeries.from_series(pc_zeros)
```

Next we train a model where `output_chunk_length == horizon`. Then we predict 24 steps ahead using both `pc` and `pc_zeros` as an input. As expected, this produces identical results since manipulating `past_covariates` in prediction window should not affect the results.

```python
model = LightGBMModel(
    lags=10, 
    lags_past_covariates=10, 
    output_chunk_length=24,
    random_state=0)

model.fit(series=y_train, past_covariates=pc_train);

y_pred = model.predict(24, past_covariates=pc)
y_pred_noise = model.predict(24, past_covariates=pc_zeros)

y_pred.plot(label='pc')
y_pred_noise.plot(label='pc_zeros')
```

    
![output_6_0](https://github.com/unit8co/darts/assets/23170686/93ad0564-6cc2-4116-b160-49f0cc4a4bda)

    

However, when `output_chunk_length < horizon`, the results are markedly different:

```python
model = LightGBMModel(
    lags=10, 
    lags_past_covariates=10, 
    output_chunk_length=1,
    random_state=0)

model.fit(series=y_train, past_covariates=pc_train);

y_pred = model.predict(24, past_covariates=pc)
y_pred_noise = model.predict(24, past_covariates=pc_zeros)

y_pred.plot(label='pc')
y_pred_noise.plot(label='pc_zeros')
```

    
![output_8_0](https://github.com/unit8co/darts/assets/23170686/1eb0f418-8dbe-43dd-a903-674c39462725)

    

## Conclusion

It seems that there is a look-ahead bias with specific LightGBM configuration. This is somewhat dangerous given that for `LightGBMModel` the `output_chunk_length` is an optional parameter defaulting to 1 and there is no warning in the current class [documentation](https://unit8co.github.io/darts/generated_api/darts.models.forecasting.lgbm.html?highlight=lightgbmmodel#darts.models.forecasting.lgbm.LightGBMModel):

> output_chunk_length (int) – Number of time steps predicted at once by the internal regression model. Does not have to equal the forecast horizon n used in predict(). However, setting output_chunk_length equal to the forecast horizon may be useful if the covariates don’t extend far enough into the future.

## Proposed solution

As a quick fix, enforce `output_chunk_length==horizon`. 

## System

- Python version: 3.8.13
- Darts version: 0.24.0

**Comment by holuond** [comment]
Hello Jalmari (@tuomijal),
I've been looking into Darts for a few weeks, so I'm far from being advanced in its inner workings and I have to say I originally shared your confusion. Yesterday I was exploring Darts documentation and gitter forum with the primary focus on understanding output_chunk_length and multi_models parameters and how they behave with LightGBM, and I think at the very least I can point you to a few relevant materials.

The primary resource I can refer you to is [this section](https://unit8co.github.io/darts/userguide/covariates.html#id6) of the User Guide, which seems highly related and why I now think the behavior your describe is actually intended, even though it might be somewhat confusing at first sight. (I can certainly say it was quite confusing for me, as the associated documentation is a bit brief in this case, which is why I had to look into the code, gitter and the associated pull requests). In the linked section, notice that:

> Time span requirements to use the same past and / or future covariates series for both fit() and predict(): with n > output_chunk_length (in your case 24 > 1): for past_covariates, at least the same time span as target plus the next n - output_chunk_length (in your case 24 - 1 = 23) time steps after the end of target

...and also:

> Depending on your forecast horizon n, the model can either predict in one go, or auto-regressively, by predicting on multiple chunks in the future. That is the reason why when predicting with past_covariates you have to supply additional “future values of your past_covariates“.

I think the auto-regression and the usage of "future values of past_covariates" when setting output_chunk_length to 1 explains the difference in the results of your two scenarios.

Secondly, I found these two gitter threads with detailed example visualizations very helpful for understanding output_chunk_length (and also multi_models, which is a related parameter one should also be aware of and understand - though it also makes the reasoning about these topics one dimension more complex):
https://matrix.to/#/!uumevxjBaNJhovFYgj:gitter.im/$yVx9kUsUHpJhRzniCIpWcKLBfxZVh2nApt32bR3bNCY?via=gitter.im&via=matrix.org&via=matrix.thegolem.cz
https://matrix.to/#/!uumevxjBaNJhovFYgj:gitter.im/$FOt0HttBLXyVe0a7XS0YgA6aG7uVnr3F3Xa-pg_XmLw?via=gitter.im&via=matrix.org&via=matrix.thegolem.cz
In the second thread (although that one has multi_models = False), notice especially the second picture (the one with the red error line), where n == 4 and ocl == 3, I think if past_covariates were used, the same logic would apply to that scenario as well and 4-3=1 past_covariates "in the future" would be used (i.e. t+1).

⚠️ This however, I just noticed, goes directly against [this Quickstart section](https://unit8co.github.io/darts/quickstart/00-quickstart.html#Covariates:-using-external-data), which says:
> past_covariates are series not necessarily known ahead of the forecast time. Those can for instance represent things that have to be measured and are not known upfront. **Models do not use the future values of past_covariates when making forecasts.**

**Comment by tuomijal** [comment]
Hi @holuond and thank you for the input. Unfortunately it does not address the issue. As you cite yourself, it should be that:

> Models do not use the future values of past_covariates when making forecasts.

This is not the case with LightGBM when `output_chunk_length<horizon` as documented in the MRE above. There is no confusion, there is a bug.

**Comment by holuond** [comment]
Yes, I think you're right @tuomijal. I was examining this in the documentation but also in the code, which might itself be the root source of the problematic behavior (and the cited part of the documentation doesn't really help to alleviate the confusion with "future values of past_covariates"). I also use only LightGBM at the moment, but this might apply to other models calling the parent `predict()` implementation in RegressionModel that use past_covariates. Let's see.

**Comment by tuomijal** [comment]
What do you think @hrzn and @dennisbader?

**Comment by holuond** [comment]
I'm also still very interested in hearing a maintainer's take on this. As far as I know hrzn no longer works on the project. I know from Gitter that @madtoinou is aware of this issue, maybe he could look into it.

**Comment by dennisbader** [comment]
Hi @tuomijal and @holuond and sorry for the late response.

Let me try to explain. In short: It is the intended behavior and applies to all GlobalForecastingModels (all RegressionModels, TorchForecastingModels, ...).


- When `horizon <= output_chunk_length`: the models do not touch future values of the past covariates because all points can be predicted in one chunk.
- When `horizon > output_chunk_length`: the model does not touch future values of the past covariates **in the first prediction chunk**. Then it moves the forecasting point `output_chunk_length` steps ahead and consumes it's previous predictions and future values of the past covariates as input for the second chunk prediction (auto-regressive). However, from the perspective of the shifted prediction point, it only uses past values of the past covariates. This continues until we reach the `horizon` *. 
- Regarding look-ahead bias: It is always a good practice to think about what data we have available at prediction time. If I do know the future values of my past_covariates ahead of time (e.g. some datetime attributes like the value of the month, day, ...) then we think it is fair to use them for predicting with `horizon > output_chunk_length`. If we do not have future values available, then we cannot predict anyway with `horizon > output_chunk_lenght`. Your example with setting the `past_covariates` to `0` is basically feeding the model with bad input -> the output will also be bad. 
For this reason we don't enforce `horizon==output_chunk_length`. 
- Instead of the auto-regressive prediction, you can also use `historical_forecasting` where the model uses the **actual** past target and covariates values for each (shifted) prediction point (if `forecast_horizon <= output_chunk_length`).

 
*We have an illustration that shows the difference between `horizon<=output_chunk_length` and `>output_chunk_length` for our TorchForecastingModels [here](https://unit8co.github.io/darts/userguide/torch_forecasting_models.html#id4). I hope this explains it better.

Also to continue on your example. If we use output_chunk_length=12, you can see that the first 12 predicted points are the same for both `past_covariates` series. This is the first predicted chunk. The next predicted chunk are the second halve of the points (blue line; it uses only past covariate values from the first halve/past/left side of the prediction point) 
```
model = LightGBMModel(
    lags=10,
    lags_past_covariates=10,
    output_chunk_length=12,
    random_state=0)

model.fit(series=y_train, past_covariates=pc_train)

y_pred = model.predict(24, past_covariates=pc)
y_pred_noise = model.predict(24, past_covariates=pc_zeros)

y_pred.plot(label='pc')
y_pred_noise.plot(label='pc_zeros')
plt.show()
```
<img width="598" alt="image" src="https://github.com/unit8co/darts/assets/84317528/3f088258-1a64-4395-9210-17f89bcf34eb">


I do agree that the section about past covariates in the QuickStart could be made more clear. Would you be interested in contributing?

**Comment by tuomijal** [comment]
Hi @dennisbader and thank you for looking into this with this level of detail.

Unfortunately, I still believe the behaviour is bad regardless if it is intended or not. If we first define that:

> past covariates are (by definition) covariates known only into the past (e.g. measurements)

and then excplicitly state that:

> Models do not use the future values of past_covariates when 
making forecasts.


we then directly violate both of these statements when we use `future values of past_covariates` in prediction window.  I don't see any reason why we should ever do so. `past_covariates` and `future_covariates` are useful and clearly defined abstractions:

![image](https://github.com/unit8co/darts/assets/23170686/8b8b0b92-4328-424a-b873-8f10e8a2692d)

Following these definitions, when you then say that:

> If I do know the future values of my past_covariates ahead of time (e.g. some datetime attributes like the value of the month, day, ...) then we think it is fair to use them for predicting with

arent those variables then by definition `future_covariates`? You give datetime attributes as an example but shouldn't these variables belong to `future_covariates` category anyways? 

I still believe this is an issue of the source code, not the documentation.

**Comment by RashidBakirov** [comment]
@dennisbader It was also my expectation that the future values of `past_covariates `should not be used for forecasting, so was surprised to read this thread.

I am wondering what happens in case that there are no future values of `past_covariates` specified, which is normally the case for real-life forecasting ahead applications. Does it mean that for this case, no auto-regressive forecasting is used, even if `output_chunk_length`<`horizon`?

**Comment by dennisbader** [comment]
Auto-regression is only used when `horizon > output_chunk_length`. Up until output_chunk_length, the model predicts directly/in one go and only uses the past values of past covariates.

If you don't have future values of your past covariates, it is not possible to predict `n > output_chunk_length`. The model raises an error in this case that your past covariates are too short. 
So in the real-life scenario it is not possible the have a look-ahead bias with these features (assuming they are not known in advance. If they are however known in advance then we can predict auto-regressively).

There are some models which do not accept **future** covariates (such as NBEATSModel, BlockRNNModel, ...). They do however accept past covariates. We can still use any potential future covariates as `past_covariates` with these models, and then perform auto-regression when `n > output_chunk_length`.

**Comment by tuomijal** [comment]
I am looking this from the user's perspective. 

I believe the user should be able to trust that the software doesn’t violate its own promises. If `past_covariates` are known in advance (i.e. in prediction range) they are then by definition `future_covariates` and should be fed to the API as such. Whatever happens under the hood (autoregression or not) is secondary.

Currently `output_chunk_length=1` by default with LightGBM. *It means that in multihorizon forecasting the user ends up using future values of `past_covariates` by default*. This happens dangerously easily:

```python
# implicit output_chunk_length=1
model = LightGBMModel(
    lags=10, 
    lags_past_covariates=10, 
    random_state=0)

model.fit(series=y_train, past_covariates=pc_train);
# biased y_pred
y_pred = model.predict(24, past_covariates=pc)
```

In other words the current design generates look-ahead bias by default.

Can we at least agree that this is an issue? If nothing else then at least the API should be modified so that `output_chunk_length` defaults to `horizon` (unless excplicitly defined) whenever `model.predict()` is called.

**Comment by madtoinou** [comment]
Hi @tuomijal,

My take on darts terminology is that `past_covariates` corresponds to information known "for the predicted timestamp", not "for the all timestamps of the forecast horizon". Moreover, the documentation explicitly details this behavior: "There are some extra nuances that might be good to know. For instance, deep learning models in Darts can (in general) forecast output_chunk_length points at a time. However it is still possible for models trained with past covariates to make forecasts for some horizon n > output_chunk_length if the past_covariates are known far enough into the future. In such cases, the forecasts are obtained by consuming future values of the past covariates, and using auto-regression on the target series" (just above [this section](https://unit8co.github.io/darts/userguide/covariates.html#introduction-what-are-covariates-in-darts))

How does this violate any promise/assumption? The model does no access `past_covariates` values beyond the predicted timestamp, hence no look-ahead bias is occurring for each timestamp composing the forecast. And if `output_chunk_length=n` (forecast_horizon), then none of the `past_covariates` beyond the first predicted timestamp will be accessed (which from what I understand, would be the behavior that you desire also when `output_chunk_length=1` with `n>1`?).

**Comment by dean-sh** [comment]
I agree with the fact that this is 1. confusing, 2. can lead to different behaviour in production than anticipated.

If I deploy such model with `output_chunk_length=1` and my horizon is, say, `24`, the model expects me to supply future values of the past_covariates. This is impossible in the same way as I cannot supply the future values of the target - that is what I'm predicting. 

In your example you "solve" this by appending **forecasts** to past_covariates.. which makes them future covariates anyway..
https://unit8co.github.io/darts/userguide/torch_forecasting_models.html#id4:~:text=temperature%20%3D%20temperature.concatenate(temperature_forecast%2C%20axis%3D0)

For now I will use `multi_models = True` and `output_chunk_length=1`, `horizon=24`. this way all of the horizons are predicted with actual past values - but I am losing the auto-regressive option.

<img width="892" alt="Screenshot 2023-08-09 at 13 54 10@2x" src="https://github.com/unit8co/darts/assets/26027302/cd66a7ac-0f3c-4b71-90c6-bc1d5748ee3a">

**Comment by madtoinou** [comment]
> If I deploy such model with `output_chunk_length=1` and my horizon is, say, `24`, the model expects me to supply future values of the past_covariates. This is impossible in the same way as I cannot supply the future values of the target - that is what I'm predicting.

I agree with you that the example provided in the documentation might be a bit counter-intuitive, however I don't think that having access to some "future values" of the past covariates them make them future covariates per se. You could have access to a limited forecast of the weather in the next 48 hours and try to predict some variables for the upcoming 7 days.

> For now I will use `multi_models = True` and `output_chunk_length=1`, `horizon=24`. this way all of the horizons are predicted with actual past values - but I am losing the auto-regressive option.

You would like the model to use the values of the past covariates up to the first predicted timestamp but then, be able to perform auto-regression on its own outputs (while always looking at the same past covariates)? So any `lags_past_covariates` would actually allow you to forecasts an arbitrary number of steps in the future? (I think that you made a typo, you probably meant `output_chunk_length=24` because otherwise, an error will be raised).

**Comment by tuomijal** [comment]
Hi @madtoinou,

Thank you for your input and apologies for the delay. 

I am afraid we are going in circles and I don't have much to add to everything I have said before. Please refer to my previous posts for context. With that in mind:

> My take on darts terminology is that past_covariates corresponds to information known "for the predicted timestamp"...

I don't think there should be any other "takes" on the terminology other than that clearly stated in the documentation:

> past covariates are (by definition) covariates known only into the past (e.g. measurements)

and

> Models do not use the future values of past_covariates when
making forecasts.

These are the "violated promises" I am referring to. They are violated when future values of `past_covariates` are nevertheless used _by default_ in backtesting (see [above](https://github.com/unit8co/darts/issues/1822#issue-1748306977)) .

I am aware that "the documentation explicitly details this behavior". What I am saying is that the documented behaviour is bad.

In reference to @dennisbader's [post](https://github.com/unit8co/darts/issues/1822#issuecomment-1649645924) above:

> If you don't have future values of your past covariates, it is not possible to predict n > output_chunk_length. The model raises an error in this case that your past covariates are too short.
So in the real-life scenario it is not possible the have a look-ahead bias with these features (assuming they are not known in advance. If they are however known in advance then we can predict auto-regressively).

If the setup does not work in real-life scenario, it should not work in backtesting either because the whole point of backtesting is to simulate the real-life as accurately as possible.

**Comment by dennisbader** [comment]
Hi @tuomijal, I give it another go

> They are violated when future values of past_covariates are nevertheless used by default in backtesting (see https://github.com/unit8co/darts/issues/1822#issue-1748306977) .

This is not true, as by default backtesting uses `forecast_horizon=1` which will always be `<= output_chunk_length` and therefore never use future values of past covariates.

> If the setup does not work in real-life scenario, it should not work in backtesting either because the whole point of backtesting is to simulate the real-life as accurately as possible.

This is indeed a look-ahead bias and can happen if the user is not aware of the relationship between forecast_horizon and output_chunk_length. However, the latest point that the user will notice that it doesn't work is when actually applying to model for the real forecasting task. Here the model will not work, and it's impossible to put it into production.

Why do we allow this then?
I mentioned it before: While for regression models (such as LightGBMModel) it might not seem to make sense as we have support for both past and future covariates, for other model like BlockRNN, NBEATSModel, NHiTSModel, TCNModel, TransformerModel we only have past covariates support. 
We still want to allow using these models in an auto-regressive fashion with forecast_horizon >= output_chunk_length. If we have future known covariates, we can still feed them to the model as past covariates to help improve forecasts.
Only because we know them in advance and they are technically future covariates, it should lock us out of feeding them to the model.
Because in Darts we aim to unify the logic of the models, we allow auto-regression for all models (even if they use past covariates).

In my opinion it is better to reformulate the description of past covariates in the documentation.

Would a warning at prediction time if the user predicts with horizon>output_chunk_length and the model uses past covariates be a fair compromise?

**Comment by tuomijal** [comment]
> Would a warning at prediction time if the user predicts with horizon>output_chunk_length and the model uses past covariates be a fair compromise?

I think this is a good idea 👍🏼. I would also consider modifying the documentation on `output_chunk_length` parameter to include a word or two on this but I leave it up to you.

> output_chunk_length (int) – Number of time steps predicted at once by the internal regression model. Does not have to equal the forecast horizon n used in predict(). However, setting output_chunk_length equal to the forecast horizon may be useful if the covariates don’t extend far enough into the future.

**Comment by dennisbader** [comment]
Sounds good @tuomijal, let me add it to our backlog.
